h2o-probes.d

```diff
@@ -23,6 +23,7 @@
 provider h2o {
     probe h1_accept(uint64_t conn_id, struct st_h2o_socket_t *sock, struct st_h2o_conn_t *conn);
     probe h1_close(uint64_t conn_id);
+    probe h2_unknown_frame_type(uint64_t conn_id, uint8_t frame_type, struct st_h2o_http2_frame_t *frame);
```

Comment:
I wonder if there is a reason to provide `frame_type` separately from `frame`. `h2o_http2_frame_t` has `frame_type` as its member.

That said, I might argue that all we need to log is the frame type. That would likely be enough to identify the feature the peer is using, and that would make this probe as simple as `probe h2o_unknown_frame_type(uint64_t conn_id, unsigned frame_type)`, and changing `uint8_t` to `unsigned` here would likely solve the CI issue that you are seeing.

Reply:
As you pointed out I also thought that the frame type would be the most important information here and most of the time users would only care about it, that's why I made it a separate integer parameter so that users can avoid decomposing structure every time. So this redundancy was intentional.

However, now I think that probably we can drop the structure parameter for now until we find a clear use case for other fields.

Interestingly, using `uintXX_t` in this file seems to be okay (we already have `uint64_t` for other functions). I confirmed the fix in local Docker. Hopefully this time it will pass.

Reply:
> I confirmed the fix in local Docker.

I take it back. So the difference was in the build mode. I have been using the Debug build whereas Docker/Travis CI use Release build. With Release build it was quite easy to reproduce the failure.

Turns out using datatypes other than `uint64_t` for arguments may not be a good idea.

https://github.com/iovisor/bpftrace/blob/master/docs/reference_guide.md#variables
> arg0, arg1, ..., argN. - Arguments to the traced function; assumed to be 64 bits wide

http://dtrace.org/guide/chp-variables.html#chp-variables-5
> int64_t arg0, ..., arg9 | The first ten input arguments to a probe represented as raw 64-bit integers.

I'll change the type for `frame_type` to `uint64_t`.

Interestingly, there are two cases where we use 32bit integer for an argument:
```
    probe h3_stream_set_state(uint64_t conn_id, uint64_t req_id, unsigned state);
    probe receive_request(uint64_t conn_id, uint64_t req_id, int http_version);
```
I'm not sure why these have been okay so far. Maybe just we have been lucky enough?

Reply:
Wow. Thank you for finding the cause, and thank you for checking other cases in which we might hit the same issue.

Could you please confirm my understanding (stated below), as I think we need to understand the impact of the problem as how to avoid hitting it?

IIUC, this is not a bug of the probing framework itself (i.e. SDT probes + bcc), but rather a bug of bpftrace that is a wrapper of bcc.

When we look into the object file generated from connection.c, we find information that correctly specifies the size of the argument (in this case `1` of `1@4(%rsp)`):
```
.pushsection .note.stapsdt,"?","note"
.balign 4
.4byte 992f-991f,994f-993f,3
991: .asciz "stapsdt"
992: .balign 4
993: .8byte 990b
.8byte _.stapsdt.base
.8byte h2o_h2_unknown_frame_type_semaphore
.asciz "h2o"
.asciz "h2_unknown_frame_type"
.asciz "8@32(%r13) 1@4(%rsp)"
994: .balign 4
.popsection
```

But because bpftrace simply ignores the size (as you point out the doc, and my read of https://github.com/iovisor/bpftrace/blob/master/src/ast/irbuilderbpf.cpp#L334-L365 confirms that), we are reading 7 bytes of unknown data.

Assuming that these understandings are correct, I tend to wonder if the long term solution would be to fix the problem in bpftrace, and that the short time solution would be to apply bit-wise AND in the probe script that we feed into bpftrace.

WDYT?

Reply:
Hmm, are you trying to go beyond the spec so that bcc/bpftrace can correctly handle integer arguments with 8,16,32-bit sizes? Not saying that's wrong, just wanted to clarify your goal first.

Reply:
Or are you suggesting that type of `argN` will remain 64bit but when bpftrace fills the value it should read the correct bit width from the process being probed and zero-clear the upper bits? That sounds good to me, and your long and short-term solutions make sense to me as well.

Reply:
Yeah I now realize that we might be conflating two different points.

What the type of argX is is something local to bpftrace. It can be 64 bits.

How bpftrace expands an argument to its internal type (i.e. 64 bit value) is a different issue, and I'd argue that it should be properly extended. I am not sure if it is possible for bpftrace (or bcc) to determine if an argument is signed or not, but at least it is capable of determining the size of the argument. Therefore, it should not incorporate junk (e.g., 0x100) when extending the argument to 64-bit.

As a side-note, it is my understanding that IRBuilder of bpftrace is only half way done; it only handles some of the most frequently observed patterns like `%reg` or `size@off(%reg)`. Depending on how you write code, you might hit a case in which your C compiler generates a pattern that cannot be handled by bpftrace.

Based on this view, my argument is that we should not change the argument types being defined in *-probes.d just because it cannot be handled properly by bpftrace. Rather, we should add workaround to our trace script, or fix bpftrace.

Reply:
Sorry I missed your comment right above. I think we are on the same page :-)

Reply:
👍 Implemented the workaround you proposed in f5849951b668b1f0b8e0f7625204eb21b99f2f0b. Will see how it goes.